### ISSUE TITLE
Micro vs. macro repositories

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo#801]
#800 started a mini-discussion in the dev meeting today about how much we want to be including in this and other Nengo-related repositories. We'll discuss this in more detail in the next dev meeting (or another upcoming one), but we can add some ideas to this issue to seed that discussion.

There are two extremes for this, and right now we're somewhere in the middle.
### Extreme 1: Microrepositories

With this approach, each repository would be responsible for the least amount of functionality possible. If a user wants access to certain functions, they may have to do a `pip install`. With this approach, we might end up with this kind of set of Python modules:
- `nengo`
- `spa`
- `nengo_gui`
- `nengo_ocl`
- `nengo_spinnaker`
- `nengo_convnet`
- `nengo_auditory`

Some of these we have now (separate modules for backends) but not others (`spa` is part of `nengo`, for example).

Pros:
- :+1: Test suites should be smaller, faster to run for each repo
- :+1: Codebase in each repo is smaller, possibly easier to read / understand
- :+1: Concerns are separated more clearly
- :+1: Each repo could develop faster as they don't care if they break a dependent repo

Cons:
- :-1: More boilerplate / duplication in repos (license text etc.)
- :-1: Repos depend on each other, and can break when dependencies update
- :-1: More cumbersome process for users / developers who use multiple repos
### Extreme 2: Macrorepositories (or macrorepository?)

With this approach, we'd put as much as is reasonable in as few repositories as possible. Users would `pip install nengo`, and they would have everything at their disposal.

Pros:
- :+1: Simple process for users / developers; even if they only use parts of it, the process is the same (`pip install` or `git clone` a single thing)
- :+1: There's little cost for extra things being installed (our scripts / binaries don't take up that much space)
- :+1: Dependent submodules stay up-to-date, as testing will identify when interactions cause parts of the whole to break

Cons:
- :-1: Test suites will be large, possibly testing many interactions between submodules, making them slow to run
- :-1: May be difficult to navigate the codebase, find where a certain class / module lives

---

Right now, we basically live somewhere in the middle. We generally will put anything that's useful and in pure Python in Nengo, except backends, and so far, custom neuron types (e.g., https://github.com/fragapanagos/adaptive_lif).

Should we start heading toward one of the two extremes, or should we continue evaluating each decision case-by-case?

**Comment by tbekolay** [comment]
Also, definitely help with adding pros and cons to either of those lists! But this might resolve faster by talking in person at the dev meeting, so no need for super long responses.

**Comment by hunse** [comment]
The reason I made this PR, and why I want this in Nengo, is that I want to put it in nengo_ocl. I can have custom builders for nengo_ocl that will do the convolution on the GPU, and let me run my convnets way faster.

What's not clear to me is how we might accommodate this with the microrepository approach. One possibility is to have micro-microrepositories, such as `nengo_ocl_convnet` to supplement `nengo_convnet`, etc. Or we could include OCL code, SpiNNaker code, etc. all in the `nengo_convnet` repository. I think the biggest downside to both these is that if backends are not written to be modular and easily extensible, then it might be very difficult to add support for these things without changing the backend itself. The reason we can even talk about the microrepository approach is that we've been very good about making the reference backend modular and extensible.

At the same time, we do want to be careful about what we put into core `nengo`. Is everything in `nengo` expected to be supported by every backend? I think we've decided that that's not necessary (based on previous conversations), but if we do put something in core `nengo` it does suggest that it's a higher priority for backend support than things which are part of microrepositories.

Obviously I'm a bit biased, since this is my PR, but I think this stuff belongs in `nengo`. We've talked for a while about having pre-built vision modules that people can easily add to their models for a while, and I think this is an important step in that direction.

Also, sorry to miss the dev meeting today!

**Comment by jgosmann** [comment]
It might be worthwhile to note that Scipy handles this stuff somewhat similar to microrepositiories with the scikits (like scikit-learn). But they have the possibility that scikits get included in the main Scipy distribution at some point if they become stable enough (and are of enough general interest).

Matplotlib also has toolkits, some of which are included with Matplotlib and some of which are not.

**Comment by bjkomer** [comment]
I like the idea of having the things that are useful to most users and in pure Python all in one place. I'm leaning toward having a couple of macrorepositories, such as one for core `nengo`, one for `nengo_gui`, and another for useful things that don't quite fit into core, but a lot of users might want (custom neuron models, utilities for building specific types of networks, etc) `nengo_utils`?. It makes sense to also have microrepositories for backends and more specialized utilities.

One of my worries if we start to split things up too much, is that testing might end up getting more complicated for the interactions between modules. For example, we might come across a bug where a particular neuron model doesn't work with some part of spa. That could be a problem with the neuron model, or spa, or a combination of both. Which repo should run this test? Do we need a separate set of 'interaction' tests? Which repo should the issue exist in?

**Comment by Seanny123** [comment]
After discussing during the meeting the consensus seems to be that newer in-development modules (such as Nengo GUI) should be maintained as separate repositories until they are stable enough to be joined into the main Nengo repository. The fact that certain modules (such as `spa`) need to be brought up to date occasionally and often lag behind the development pace of the main repository is considered acceptable compared to the inconvenience of keeping track of things in a bunch of different places.

**tl;dr** we're going with the macro-repository model

P.S. should this issue be closed now?

**Comment by tcstewar** [comment]
That sounds about right to me; although I'm undecided as to whether we'll ever end up merging ```nengo_gui``` or ```nengo_ocl``` into ```nengo``` even if they do become nice and stable.

That said, I think both ```spa``` and all the backend helper functions should be in ```nengo```, as we currently do it.  The main drawback is that we do have to be fairly careful about maintaining backwards compatibility (or quickly updating the backends that get broken).  We saw that recently with the change to the cache system, which broke ```nengo_spinnaker```.  But it was a really easy fix -- it was just a matter of doing it.  But I think similar things are going to happen with just about any change to ```nengo```, so it's just something to deal with and I don't think there's any solution that doesn't involve merging *everything* into one repository, which I don't think is a good idea.

**Comment by tbekolay** [comment]
Let's keep this open -- I want to add it to the developer documentation. Also made a cute label for issues this like. :octopus:

**Comment by Seanny123** [comment]
> I'm undecided as to whether we'll ever end up merging nengo_gui or nengo_ocl into nengo even if they > do become nice and stable.

One of the things we talked about in the meeting is always keeping the back-ends separate from the main repository.

**Comment by tcstewar** [comment]
> One of the things we talked about in the meeting is always keeping the back-ends separate from the main repository.

Definitely.  And I believe the consensus was that all of the non-default backends would be kept separate.  But the reference backend would stay in ```nengo```.

**Comment by tbekolay** [comment]
The `document this` label is now valid on closed issues, so I'm closing this even though there may be things left to document.
